enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/busted/BaseBustedRecordFormat.java

```diff
@@ -0,0 +1,244 @@
+ * on both and potentially re-reading the second or both until a consistent read was had.
+ *
+ * @param <RECORD> type of {@link AbstractBaseRecord}
+ */
+public abstract class BaseBustedRecordFormat<RECORD extends AbstractBaseRecord>
+        extends BaseOneByteHeaderRecordFormat<RECORD>
+{
+    static final long NULL = Record.NULL_REFERENCE.intValue();
+    static final int HEADER_BIT_RECORD_UNIT = 0b0000_0010;
+    static final int HEADER_BIT_FIRST_RECORD_UNIT = 0b0000_0100;
+
+    protected BaseBustedRecordFormat( Function<StoreHeader,Integer> recordSize, int recordHeaderSize )
+    {
+        super( recordSize, recordHeaderSize, IN_USE_BIT );
+    }
+
+    @Override
+    protected final void doRead( RECORD record, PageCursor primaryCursor, int recordSize, PagedFile storeFile,
+            long headerByte, boolean inUse ) throws IOException
+    {
+        boolean recordUnit = has( headerByte, HEADER_BIT_RECORD_UNIT );
+        if ( recordUnit )
+        {
+            boolean firstRecordUnit = has( headerByte, HEADER_BIT_FIRST_RECORD_UNIT);
+            if ( !firstRecordUnit )
+            {
+                // This is a record unit and not even the first one, so you cannot go here directly and read it,
+                // it may only be read as part of reading the primary unit.
+                record.clear();
+                return;
+            }
+        }
+
+        long secondaryId = -1;
+        DataAdapter<PageCursor> dataAdapter = PAGE_CURSOR_ADAPTER;
+        SecondaryPageCursorControl secondaryPageCursorControl = SecondaryPageCursorControl.NULL;
+        if ( recordUnit )
+        {
+            int primaryEndOffset = primaryCursor.getOffset() + recordSize - 1 /*we've already read the header byte*/;
+
+            // This is a record that is split into multiple record units. We need a bit more clever
+            // data structures here. For the time being this means instantiating one object,
+            // but the trade-off is a great reduction in complexity.
+            secondaryId = Reference.decode( primaryCursor, dataAdapter );
+            @SuppressWarnings( "resource" )
+            SecondaryPageCursorReadDataAdapter readAdapter = new SecondaryPageCursorReadDataAdapter(
+                    primaryCursor, storeFile,
+                    pageIdForRecord( secondaryId, storeFile.pageSize(), recordSize ),
+                    offsetForId( secondaryId, storeFile.pageSize(), recordSize ),
+                    primaryEndOffset, PagedFile.PF_SHARED_READ_LOCK );
+            dataAdapter = readAdapter;
+            secondaryPageCursorControl = readAdapter;
+        }
+
+        try
+        {
+            do
+            {
+                // (re)sets offsets for both cursors
+                secondaryPageCursorControl.reposition();
+                doReadInternal( record, primaryCursor, recordSize, headerByte, inUse, dataAdapter );
+            }
+            while ( secondaryPageCursorControl.shouldRetry() );
```

Comment:
in case when `!recordUnit` secondaryPageCursorControl will always be SecondaryPageCursorControl.NULL with empty `shouldRetry`. Should we still retry primaryCursor here?